cpp/src/arrow/error_or.h

```diff
@@ -0,0 +1,427 @@
+      // ErrorOr-specific error message.
+      other.variant_.status_ =
+          Status::Invalid(ErrorOrConstants::kStatusMoveAssignmentMsg);
+#endif
+    }
+
+    return *this;
+  }
+
+  /// Indicates whether the object contains a `T` value.
+  ///
+  /// \return True if this ErrorOr object's status is OK (i.e. a call to ok()
+  /// returns true). If this function returns true, then it is safe to access
+  /// the wrapped element through a call to ValueOrDie().
+  bool ok() const { return has_value_; }
+
+  /// Gets the stored status object, or an OK status if a `T` value is stored.
+  ///
+  /// \return The stored non-OK status object, or an OK status if this object
+  ///         has a value.
+  Status status() const { return ok() ? Status::OK() : variant_.status_; }
+
+  /// Gets the stored `T` value.
+  ///
+  /// This method should only be called if this ErrorOr object's status is OK
+  /// (i.e. a call to ok() returns true), otherwise this call will abort.
+  ///
+  /// \return The stored `T` value.
+  const T& ValueOrDie() const& {
+    if (!ok()) {
+      std::abort();
+    }
+    return variant_.value_;
+  }
+
+  /// Gets a mutable reference to the stored `T` value.
+  ///
+  /// This method should only be called if this ErrorOr object's status is OK
+  /// (i.e. a call to ok() returns true), otherwise this call will abort.
+  ///
+  /// \return The stored `T` value.
+  T& ValueOrDie() & {
+    if (!ok()) {
+      std::abort();
+    }
+    return variant_.value_;
+  }
+
+  /// Moves and returns the internally-stored `T` value.
+  ///
+  /// This method should only be called if this ErrorOr object's status is OK
+  /// (i.e. a call to ok() returns true), otherwise this call will abort. The
+  /// ErrorOr object is invalidated after this call and will be updated to
+  /// contain a non-OK status.
+  ///
+  /// \return The stored `T` value.
+  T ValueOrDie() && {
+    if (!ok()) {
+      std::abort();
+    }
+    T tmp(std::move(variant_.value_));
+
+    // Invalidate this ErrorOr object.
+    OverwriteValueWithStatus(Status::Invalid(ErrorOrConstants::kValueOrDieMovedMsg));
+    return std::move(tmp);
+  }
+
+ private:
+  // Resets the |variant_| member to contain |status|.
+  template <class U>
+  void AssignStatus(U&& status) {
+    if (ok()) {
+      OverwriteValueWithStatus(std::forward<U>(status));
+    } else {
+      // Reuse the existing Status object. has_value_ is already false.
+      variant_.status_ = std::forward<U>(status);
+    }
+  }
+
+  // Under the assumption that |this| is currently holding a value, resets the
+  // |variant_| member to contain |status| and sets |has_value_| to indicate
+  // that |this| does not have a value. Destroys the existing |variant_| member.
+  template <class U>
+  void OverwriteValueWithStatus(U&& status) {
+#ifndef NDEBUG
+    if (!ok()) {
+      std::abort();
+    }
+#endif
+    variant_.value_.~T();
+    new (&variant_) variant(std::forward<U>(status));
```

Comment:
I'd put a comment.

Reply:
I assume you meant on the strange usage of "new", I think I covered most places where this happens.

Reply:
Yes, that this is a placement new paired with a move, it's not trivial for those who are familiar with C++.